Allow to update multiple data points in one step

DerAndereAndi opened the pull request:
Setting multiple data points, e.g. measurements for multiple phases, or setting currents and power values, should be done in one SPINE update call. Otherwise every single update will trigger an individual SPINE notification which is not nice to have.

This change updates the LoadControl and Measurement server features to allow providing multiple data points with individual filters for each, in a single call. This will then update the storage only once and trigger only one update notification.

DerAndereAndi committed:
Allow to update multiple data points in one step

Setting multiple data points, e.g. measurements for multiple phases, or setting currents and power values, should be done in one SPINE update call. Otherwise every single update will trigger an individual SPINE notification which is not nice to have.

This change updates the LoadControl and Measurement server features to allow providing multiple data points with individual filters for each, in a single call. This will then update the storage only once and trigger only one update notification.

DerAndereAndi commented:
@sthelen-enqs this is a suggestion on how to solve the issue described in the description and useful for your MPC draft PR. What do you think?

DerAndereAndi committed:
Fix gosec issues

DerAndereAndi commented on features/server/loadcontrol.go:
```diff
@@ -71,48 +71,73 @@
+	for _, item := range data {
+		filterData = append(filterData, api.LoadControlLimitDataForFilter{
+			Data:   item.Data,
+			Filter: model.LoadControlLimitDescriptionDataType{LimitId: &item.Id},
```
Fixed

DerAndereAndi commented on features/server/measurement.go:
```diff
@@ -73,48 +73,73 @@
+	for _, item := range data {
+		filterData = append(filterData, api.MeasurementDataForFilter{
+			Data:   item.Data,
+			Filter: model.MeasurementDescriptionDataType{MeasurementId: &item.Id},
```
Fixed

DAMEK86 commented on api/featuresserver.go:
```diff
@@ -90,26 +100,38 @@ type LoadControlServerInterface interface {
-		data model.LoadControlLimitDataType,
+	UpdateLimitDataForIds(
+		data []LoadControlLimitDataForID,
+		deleteId *model.LoadControlLimitIdType,
```
Why the Id is called `deleteId` when u have a method to update the limit?
It feels like the method name and the id doesn't match to each other

DerAndereAndi replied:
On each update, you can also request to delete some data. So the `data` structure contains everything you want to set, and `deleteId` and `deleteElements` references what you want to delete.

DAMEK86 replied:
But shouldn't that be an delete instead of an update method?
It's confusing why I need to pass a delete ID for an CRUD method named update..

DerAndereAndi replied:
It's to allow deletion of a field item, e.g. duration in LPC. The `data` structure doesn't allow you that. Having separate methods would requiring to introduce transactions, which I want to avoid.

The reasoning behind all that is how EEBUS works. See specs about partial updates.

DerAndereAndi replied:
One addition: the complete updates are always partial and only one delete element can be done on a single SPINE message. That's why the `deleteId` and `deleteElements` are not part of the `data` structure.

DerAndereAndi committed:
Simplify and add ElectricalConnectionPermitted

- Remove delete option from updates via ID, use update via filter if needed
- Add support for ElectricalConnectionPermittedValueSet updates

DerAndereAndi committed:
Attempt to fix data race in test

sthelen-enqs commented:
> @sthelen-enqs this is a suggestion on how to solve the issue described in the description and useful for your MPC draft PR. What do you think?

Looks decent, I can try using it in the MPC draft PR and seeing how it works out in practice.

DerAndereAndi committed:
Add tests and fix bug

- Added tests for multi data update
- Fixed a bug in ElectricalConnection update

DerAndereAndi committed:
Merge branch 'dev' into feature/update-multiple-datapoints

# Conflicts:
#	usecases/cs/lpc/usecase_test.go
